jujugui/static/gui/src/app/app.js

```diff
@@ -469,16 +469,7 @@ YUI.add('juju-gui', function(Y) {
             serviceName: 'jem'
           });
           this.jem = new window.jujulib.environment(this.get('jemUrl'), bakery);
-
-          // XXX j.c.sackett 2015-10-07 This is only for demo to make sure the
-          // JEM is hooked up properly--we should remove it down the line.
-          this.jem.listEnvironments(function(data) {
-            console.log('Environment listing success!');
-            console.log(data);
-          }, function(error) {
-            console.log('Environment listing failure.');
-            console.log(error);
-          });
+          this._connectJEMSocketUrl();
```

Comment:
This cannot be called here.
At the time this function is called, we still don't have this.env, required by the `this.jem.listEnvironments` callback in `this._connectJEMSocketUrl()` so we are relying on good timing. We need to first get the env list from JEM, then calculate the initial socket_url and finally instantiate the GoEnvironment (this.env). The remaining app intialization needs to be deferred, the code must reflect that we now depend on JEM, but we also want to fall back gracefully if JEM is not there (so if `this.get('jemUrl')` is not defined?), for plan 9. I suspect that `this._generateSocketUrl()` would be a good place for this logic. I am quite convinced that this needs to be fixed before ODS, because we are encountering weird state issues and setCredentials races, for instance when you refresh after switching to another environment and the login form is displayed.